bottlecap/src/metrics/dogstatsd.rs

```diff
@@ -35,63 +56,115 @@ impl DogStatsD {
         DogStatsD {
             cancel_token,
             aggregator,
-            socket,
-            event_bus,
+            buffer_reader: BufferReader::UdpSocketReader(socket),
         }
     }
 
     pub async fn spin(self) {
-        loop {
-            // TODO(astuyve) this should be dynamic
-            let mut buf = [0; 1024]; // todo, do we want to make this dynamic? (not sure)
-            let (amt, src) = self
-                .socket
-                .recv_from(&mut buf)
-                .await
-                .expect("didn't receive data");
-            let buf = &mut buf[..amt];
-            let msg = std::str::from_utf8(buf).expect("couldn't parse as string");
-            debug!(
-                "received message: {} from {}, sending it to the bus",
-                msg, src
-            );
-            let parsed_metric = match Metric::parse(msg) {
-                Ok(parsed_metric) => {
-                    debug!("parsed metric: {:?}", parsed_metric);
-                    parsed_metric
-                }
+        let mut spin_cancelled = false;
+        while !spin_cancelled {
+            self.consume_statsd().await;
+            spin_cancelled = self.cancel_token.is_cancelled();
+        }
+    }
+
+    async fn consume_statsd(&self) {
+        let (buf, src) = self
+            .buffer_reader
+            .read()
+            .await
+            .expect("didn't receive data");
+        let msgs = std::str::from_utf8(&buf).expect("couldn't parse as string");
+        debug!(
+            "received message: {} from {}, sending it to the bus",
+            msgs, src
+        );
+        let statsd_metric_strings = msgs.split('\n');
+        self.insert_metrics(statsd_metric_strings);
+    }
+
+    fn insert_metrics(&self, msg: Split<char>) {
+        let all_valid_metrics: Vec<Metric> = msg
+            .filter_map(|m| match Metric::parse(m) {
+                Ok(metric) => Some(metric),
                 Err(e) => {
-                    error!("failed to parse metric: {:?}\n message: {:?}", msg, e);
-                    continue;
+                    error!("failed to parse metric: {}", e);
```

Comment:
Should we also add the message in the error? so we know which metric we failed to parse and from where?

Reply:
good idea!